### DISCUSSION TITLE
Замена значений в списке

### DISCUSSION THREAD

**Discussion by mihail-cornev** [totumonline/totum-ru-issues-and-discussions#573] [Как это сделать?] [answered]
ListReplace для меня пока сложный для понимания инструмент, и документации недостаточно чтобы с ним разобраться. Может кто-то помочь с пояснениями?

Есть поле #pos и в нём значение типа список [12,215,55].
По изменению в этом поле код действия должен заменять Id из этого списка, на #nr_name из другой таблцы.
в итоге должен получится список ["12 - Перегородка", "215 - Дверь", "55 - Окно"]

т.е. через listreplace я думал заменить значения на названия, но не могу понять как это работает:

![Screenshot_15](https://github.com/totumonline/totum-ru-issues-and-discussions/assets/113771444/1e4dd859-4806-41ba-b837-b1c3b7508955)

`= : set(table: $#ntn; field: 'price' = $sel; where: 'id' = #id; log: true)`
`		sel: listReplace(list: #positions; action: "item" = $val; action: $val; value: "value")`
`				val: select(table: 'test_projects_'; field: 'nr_name'; where: 'id' = $#value)`

Подскажите что я делаю не так, и как в принципе работает listreplace. Зачем там key и value?
`=: listReplace(list: ; action: "item" = $val; action: $val; key: "key"; value: "value")`
Спасибо.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
listReplace проще чем все думают. Он обрабатывает указанный в параметре `list` список от начала до конца. Если зашел список 5 значений на выходе будет список 5 значений.

Каким образом меняются значения. Они меняются в параметрах `action`.

Если на входе у нас список `[1,5,16,4]` (например это какие-то id и нам нужно им найти соответсвующие имена), то будет так: `action: $value` и в строке `value` уже будет поиск этого артикула для каждой строчки.

Пример:

```
sel: listReplace(list: json`[1,5,16,4]`; action: $val; value: "row")
				val: select(table: 'test_projects_'; field: 'nr_name'; where: 'id' = ?)
```

Я пока оставил в сравнении `where` вопросительный знак, что бы пояснить параметр `value`. Дело в том, что когда мы в строке `val` выполняем поиск, то нам надо сравнивать со стартовым значением, которое у нас есть в обрабатываемом списке. Это значение передается в переменную, название которой мы задаем в параметре `value`. Почему мы ее задаем сами, потому что у нас в одном коде может быть 10 разных listReplace и у них должны быть разные переменные. Переменная вызывается как обычно $#value_name. В моем примере `$#row`.

Я назвал ее `row` тк в реальности в эту переменную по очереди приходят значения каждой обрабатываемой строки в том порядке, в котором они в списке источнике.

Те в моем примере на шаге 2 (нумерация начинается с 0): `$#row` = `16`.

А что будет, если у нас на входе не `list` а `rowlist`. В этом случае там будет вся строка, например: `{"name":"Alex","id":10}`

Тогда если у нас на входе `list` сравнение будет `where: 'id' = $#row`, а если обрабатываем `rowlist` то `where: 'id' = $#row[id]`.

Пример для `list`:

```
replace: listReplace(list: json`[1,5,16,4]`; action: $val; value: "row")
				val: select(table: 'test_projects_'; field: 'nr_name'; where: 'id' = $#row)
```

Теперь посмотрим что такое `action: "item" = $val` — эта штука для обработки `rowlist`. Те когда у нас на входе `rowlist` и мы хотим заменить значение только в одной его колонке или добавить новую колонку то вот тогда используется этот вариант.

Пример c `rowlist`:

```
replace: listReplace(list: #some_rowlist; action: "second_name" = $val; value: "row")
				val: select(table: 'test_projects_'; field: 'nr_name'; where: 'id' = $#row[id])
```

В этом примере он мне в каждой строке моего rowlist добавит колонку `second_name`: `{"name":"Alex","id":10; "second_name":"Some found name"}` этот second_name ищет для каждой строки по ключу `id`

Те `action: $val` — это замена всего значения строки на новое, а `action: "second_name" = $val` это изменение одной колонки или добавление новой. Второе понятно работает только если на входе `rowlist`. Поэтому ваш пример и ругается, что на самом первом шаге он пытается добавить колонку `item`, а добавить он ее не может тк на входе `list` (у листа колонок нет).

Если же на входе rowlist то можно несколько `action: "item_1" = $val_1; action: "item_2" = $val_2`.

А можно ли для входящего list несколько `action: $val`. Можно, но только в кодах-действия, потому что в кодах действия это будут выполняемые действия. Если просто в код, то это получается абсурд, тк вы последующим action перезаписываете предыдущий.

Что такое `key`? Если в `value` передается значение обрабатываемой на шаге строки, то в `key` передается ее номер (нумерация списков начинается с 0).  Для  списка `[1,5,16,4]` при обработке сроки со значением 16 в переменной определенной в `key` будет ... `2`.

Зачем это нафиг нужно? Дело в том, что в `list` можно скормить не только `list` и `rowlist`, но и `row`.

Например у нас есть такой `row`: `{"490":"Alex","520":"Mike","530":"John"}` и мы хотим для всех у кого ключи больше 500 добавить * (например это количество баллов системы лояльности):

```
replace: listReplace(list: json`{"490":"Alex","520":"Mike","530":"John"}`; action: $val; key: "key"; value: "value")
				val: if(condition: $#key > 500; then: str`$#value ++ "*"`; else: $#value)
```

Обратите внимание, что новое значение присваивается на каждом шаге — listReplace ничего не пропускает, но если нам не нужно менять значение на шаге, то мы можем новое присвоить без изменения из $#value!

Вот и все!

**Reply by ttm-alexeypolunin** [reply]
Конкретно в этом примере не нужен listReplace если на входе `[1,45,67]` а на выходе `["Окно","Батарея","Холодильник"]`.

Это делается простым selectList: 

```
sel: selectList(table: 'test_projects_'; field: 'nr_name'; where: 'id' = #positions)
```

Если же там есть какие-то условия промежуточные, тогда да. Или если нам надо получить составное значение `["1/Окно","45/Батарея","67/Холодильник"]` — в этих случаях listReplace.
